Unify cpp_extension build directory removal

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135402
* #137251
* #136695
* #136735
* __->__ #136059

Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

xmfan committed:
[compiled autograd] fix flaky cpp extensions tests

[ghstack-poisoned]

xmfan committed:
Update on "[compiled autograd] fix flaky cpp extensions tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "Unify cpp_extension build directory removal and fix flaky tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685


Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

ezyang commented on torch/utils/cpp_extension.py:
```diff
@@ -325,6 +325,18 @@ def get_default_build_root() -> str:
+        if IS_WINDOWS:
+            # rmtree returns permission error: [WinError 5] Access is denied
+            # on Windows, this is a workaround
+            subprocess.run(["rm", "-rf", default_build_root], stdout=subprocess.PIPE)
```
Does this actually work lol

albanD commented on docs/source/conf.py:
```diff
@@ -1898,6 +1898,7 @@
     "check_compiler_ok_for_platform",
     "get_cxx_compiler",
     "get_default_build_root",
+    "remove_default_build_root",
```
Please make sure to document newly added APIs by adding the appropriate entry in the .rst file+docstring or make them private.

albanD commented on torch/utils/cpp_extension.py:
```diff
@@ -325,6 +325,18 @@ def get_default_build_root() -> str:
     """
     return os.path.realpath(torch._appdirs.user_cache_dir(appname='torch_extensions'))
 
+def remove_default_build_root():
```
I'm not sure we really want a public API that rm -rf a folder like this.
It makes sense as a utility for our testing framework. But I don't think we should do it here.

albanD commented on test/inductor/test_extension_backend.py:
```diff
@@ -50,12 +49,7 @@
-        return
-    default_build_root = torch.utils.cpp_extension.get_default_build_root()
-    if os.path.exists(default_build_root):
-        shutil.rmtree(default_build_root, ignore_errors=True)
+    torch.utils.cpp_extension.remove_default_build_root()
```
You can use `torch/testing/_internal/*` to host this kind of utilities.
Also I would suggest remove this indirection since it just calls down the same function?

xmfan committed:
Update on "Unify cpp_extension build directory removal and fix flaky tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685


Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "Unify cpp_extension build directory removal and fix flaky tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685


Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "Unify cpp_extension build directory removal and fix flaky tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685


Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "Unify cpp_extension build directory removal and fix flaky tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685


Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "Unify cpp_extension build directory removal and fix flaky tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685


Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "Unify cpp_extension build directory removal and fix flaky tests"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685


Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "Unify cpp_extension build directory removal"

FIXES #125579 #131103 #133197 #133283 #134738 #135369 #135685


Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan commented:
There still seems to be tests that rely on the c++ extensions living across tests. I am opting for a different approach to fix custom ops flaky tests. This PR is now just a utils refactor

xmfan committed:
Update on "Unify cpp_extension build directory removal"



Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

xmfan committed:
Update on "Unify cpp_extension build directory removal"



Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

albanD reviewed (approved):
Ok!

xmfan committed:
Update on "Unify cpp_extension build directory removal"



Keeps existing default directory clearing logic, even though it fails when TORCH_EXTENSIONS_DIR is set. To properly clear, we'd need to track all the folders we compiled the extensions to.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]
